pkg/rke2/rke2.go

```diff
@@ -166,5 +176,145 @@ func setup(clx *cli.Context, cfg Config) error {
+			if _, err := os.Stat(manifestName); err == nil {
+				kubeletStandAlone = true
+				if err := os.Remove(manifestName); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	if kubeletStandAlone {
+		kubeletCmd := exec.Command("kubelet")
+		containerdCmd := exec.Command("containerd")
+
+		tmpAddress := filepath.Join(os.TempDir(), "containerd.sock")
+
+		// start containerd
+		go startContainerd(ctx, dataDir, containerdErr, tmpAddress, containerdCmd)
+		// start kubelet
+		go startKubelet(ctx, dataDir, kubeletErr, tmpAddress, kubeletCmd)
+
+		// check for any running containers from the disabled items list
+		go checkForRunningContainers(ctx, tmpAddress, disabledItems, kubeletCmd, containerdCmd)
+
+		go func() {
+			logrus.Infof("containerd 4 %v", containerdCmd)
+			select {
+			case err := <-kubeletErr:
+				logrus.Infof("kubelet Exited: %v, exiting Containerd", err)
+				// exits containerd if kubelet exits
+				containerdCmd.Process.Kill()
+			case err := <-containerdErr:
+				logrus.Infof("Containerd Exited: %v, exiting kubelet", err)
+				kubeletCmd.Process.Kill()
+			case <-time.After(5 * time.Minute):
+				// kill both temp process after 5 minutes anyway
+				kubeletCmd.Process.Kill()
+				containerdCmd.Process.Kill()
+			}
+		}()
+	}
+
 	return nil
 }
+
+func startKubelet(ctx context.Context, dataDir string, errChan chan error, tmpAddress string, cmd *exec.Cmd) {
+	args := []string{
+		"--fail-swap-on=false",
+		"--container-runtime=remote",
+		"--containerd=" + tmpAddress,
+		"--container-runtime-endpoint=unix://" + tmpAddress,
+		"--pod-manifest-path=" + podManifestsDir(dataDir),
+	}
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Stdout = nil
+	cmd.Stderr = nil
+
+	errChan <- cmd.Run()
+}
+
+func startContainerd(ctx context.Context, dataDir string, errChan chan error, tmpAddress string, cmd *exec.Cmd) {
+	// starting containerd first then kubelet
+	args := []string{
+		"-a", tmpAddress,
+		"--root", filepath.Join(dataDir, "agent", "containerd"),
+	}
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Stdout = nil
+	cmd.Stderr = nil
+
+	errChan <- cmd.Run()
+}
+
+func isContainerRunning(ctx context.Context, name string, address string) (bool, error) {
+	conn, err := containerdk3s.CriConnection(ctx, address)
+	if err != nil {
+		return false, err
+	}
+	c := runtimeapi.NewRuntimeServiceClient(conn)
+	defer conn.Close()
+	resp, err := c.ListContainers(ctx, &runtimeapi.ListContainersRequest{})
+	if err != nil {
+		return false, err
+	}
+
+	for _, c := range resp.Containers {
+		if c.Metadata.Name == name {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func checkForRunningContainers(ctx context.Context, tmpAddress string, disabledItems map[string]bool, kubeletCmd, containerdCmd *exec.Cmd) {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	for range ticker.C {
+		var gc bool
+		for item, disabled := range disabledItems {
+			if disabled {
+				isRunning, err := isContainerRunning(ctx, item, tmpAddress)
```

Comment:
Each call to isContainerRunning results in a ListContainers call. While not especially expensive, it seems like it would be more efficient just call ListContainers once per ticker iteration, and then look at all the names in the list and see if they're in disabledItems.

Reply:
fixing